src/solver/lu_decomposition.cu

```diff
@@ -0,0 +1,177 @@
+#include <iostream> 
+#include <vector>
+#include <micm/util/cuda_param.hpp> 
+const size_t BLOCK_SIZE = 320; 
+struct decomposeDevice{
+    double* A; 
+    double* L; 
+    double* U; 
+    char* do_aik; 
+    size_t* aik; 
+    char* do_aki;
+    size_t* aki;  
+    size_t* uii; 
+    std::pair<size_t,size_t>* niLU;
+    std::pair<size_t, size_t>* uik_nkj; 
+    std::pair<size_t, size_t>* lij_ujk;
+    std::pair<size_t, size_t>* lki_nkj; 
+    std::pair<size_t, size_t>* lkj_uji;
+}; 
+namespace micm{
+    namespace cuda{        
+        __global__ void DecomposeKernel(
+            decomposeDevice* device,
+            size_t n_grids,
+            size_t niLU_size)
+        {
+            size_t tid = blockIdx.x * blockDim.x + threadIdx.x;
+            double* A = device->A; 
+            double* L = device->L;
+            double* U = device->U;
+            std::pair<size_t, size_t>* lkj_uji = device->lkj_uji;
+            std::pair<size_t, size_t>* uik_nkj = device->uik_nkj;
+            std::pair<size_t, size_t>* lij_ujk = device->lij_ujk;
+            std::pair<size_t, size_t>* lki_nkj = device->lki_nkj;
+            size_t do_aik_offset = 0; //boolean vector 
+            size_t aik_offset = 0;
+            size_t uik_nkj_offset = 0; 
+            size_t lij_ujk_offset = 0; 
+            size_t do_aki_offset = 0; //boolean vector 
+            size_t aki_offset = 0; 
+            size_t lki_nkj_offset = 0; 
+            size_t lkj_uji_offset = 0; 
+            size_t uii_offset = 0; 
+            
+            if (tid < n_grids){
+                //loop through every element in niLU 
+                for (size_t i = 0; i < niLU_size; i++){
+                    //upper triangular matrix 
+                    auto inLU = device->niLU[i]; 
+                    for (size_t iU = 0; iU < inLU.second; ++iU){
+                        if(device->do_aik[do_aik_offset++]){
```

Comment:
Is `do_aik` here the type of `char`? If so, shall we use something like `device->do_aik[do_aik_offset++] == 'true'`?

Reply:
`do_aik[do_aik_offset]` is a single `char`, not a string. `true` will be stored as `1` and `false` as `0` (or really the ASCII character whose index is 1 or 0). It's ok to use as it is written here.

Reply:
`do_aik[do_aik_offset]` is a single `char`, not a string. `true` will be stored as `1` and `false` as `0` (or really the ASCII character whose index is 1 or 0). It's ok to use as it is written here.

Reply:
Thanks for the clarification.